File: app/android-interface/app/src/main/java/com/aburv/takbuff/data/LoginData.kt
package com.aburv.takbuff.data

import android.content.Context
import android.util.Log
import com.aburv.takbuff.db.AppUser
import com.aburv.takbuff.db.AppUserToken
import com.aburv.takbuff.db.DeviceDB
import com.aburv.takbuff.db.UserDB
import com.aburv.takbuff.db.UserTokenDB
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject

class Auth(private val context: Context) {
    companion object {
        private const val TAG = "App-LoginData"
        private const val NAMESPACE: String = "auth/"
    }

    suspend fun login(
        name: String,
        picUrl: String,
        email: String,
        googleId: String,
        location: Pair<String, String>?,
        response: LoginResponse
    ) {
        Log.i(TAG, "Login Call")
        val deviceDb = DeviceDB(context)
        val userTokenDb = UserTokenDB(context)
        val userDb = UserDB(context)
        val path = "${NAMESPACE}login"

        val locationData = JSONObject()
        if (location != null) {
            locationData.put("lat", location.first)
            locationData.put("long", location.second)
        }
        var deviceId = ""
        val devices = deviceDb.getDevices()
        if (devices.isNotEmpty()) {
            deviceId = devices[0].appDeviceId
            Log.i("Device done", deviceId)
        }
        Log.i("Device", deviceId)
        val loginData = JSONObject()

        loginData.put("deviceId", deviceId)
        loginData.put("location", locationData)

        val uIdData = JSONObject()
        uIdData.put("gId", googleId)
        uIdData.put("value", email)
        uIdData.put("type", "M")

        val userData = JSONObject()
        userData.put("name", name)
        userData.put("picurl", picUrl)
        userData.put("uId", uIdData)

        val data = JSONObject()

        data.put("user", userData)
        data.put("login", loginData)
        Log.i("Data", data.toString())

        try {
            APIRequest(context).post(path, data, object : Response {
                override fun onData(value: String) {
                    Log.i(TAG, "Received Tokens")

                    val resData = JSONObject(value)

                    val idToken = resData.getString("idToken")
                    val accessToken = resData.getString("accessToken")

                    val appUserToken = AppUserToken(
                        isActive = true,
                        accessToken = accessToken,
                        idToken = idToken,
                    )
                    val idPayload: JSONObject? = AuthUtil.parseToken(idToken)

                    if (idPayload != null) {
                        val userId = AuthUtil.getData(idPayload, "user.id") ?: ""
                        val userDp = AuthUtil.getData(idPayload, "user.dp") ?: ""
                        val userName = AuthUtil.getData(idPayload, "user.name") ?: ""
                        val userEmail = AuthUtil.getData(idPayload, "user.user_id.val") ?: ""

                        val appUser = AppUser(
                            isActive = true,
                            id = userId,
                            email = userEmail,
                            dp = userDp,
                            name = userName
                        )

                        CoroutineScope(Dispatchers.Default).launch {
                            userTokenDb.insertUserToken(appUserToken)
                            userDb.insertUser(appUser)
                        }

                        val accessPayload: JSONObject? = AuthUtil.parseToken(accessToken)
                        val isNew = isNewToApp(accessPayload!!)
                        if (isNew)
                            response.onNewUser()
                        else
                            response.onExistingUser()
                    }
                }

                override fun onError(value: String) {
                    Log.e("Login Call", value)
                    response.onError(value)
                }
            })
        } catch (e: Exception) {
            Log.i(TAG, "Failed in get auth info: $e")
        }
    }

    private fun isNewToApp(accessPayload: JSONObject): Boolean {
        val accesses: JSONArray = accessPayload.getJSONArray("accesses")
        for (i in 0 until accesses.length()) {
            val accessValue = accesses.get(i) as JSONObject

            for (key in accessValue.keys()) {
                if (key == "object_type") {
                    val value = accessValue.getString(key)
                    if (value == "U") {
                        return false
                    }
                }
            }
        }
        return true
    }

    suspend fun refreshToken() {
        Log.i(TAG, "Login Call")
        val path = "${NAMESPACE}refresh_token"
        try {
            APIRequest(context).get(path, object : Response {
                override fun onData(value: String) {
                    val resData = JSONObject(value)

                    val idToken = resData.getString("idToken")
                    val accessToken = resData.getString("accessToken")
                    val userTokenDb = UserTokenDB(context)
                    val appUserToken = AppUserToken(
                        isActive = true,
                        accessToken = accessToken,
                        idToken = idToken,
                    )
                    CoroutineScope(Dispatchers.Default).launch {
                        userTokenDb.updateUserToken(appUserToken)
                    }
                }

                override fun onError(value: String) {
                }
            })
        } catch (e: Exception) {
            Log.i(TAG, "Refresh $e.stackTraceToString()")
        }
    }
}

interface LoginResponse {
    fun onNewUser()
    fun onExistingUser()
    fun onError(message: String)
}

object AuthUtil {
    fun parseToken(idToken: String): JSONObject? {
        val parts = idToken.split(".")
        if (parts.size != 3) return null

        try {
            val payload = String(android.util.Base64.decode(parts[1], android.util.Base64.URL_SAFE))
            Log.i("Parse", payload)
            return JSONObject(payload)
        } catch (e: Exception) {
            Log.i("Parse token", e.toString())
            return null
        }
    }

    fun getData(data: JSONObject, key: String): String? {
        val keys = key.split(".")
        var current: JSONObject = data

        for ((index, currentKey) in keys.withIndex()) {
            if (index == keys.lastIndex) {
                return current.optString(currentKey)
            }
            current = current.optJSONObject(currentKey) ?: return null
        }
        return null
    }
}

File: app/i-interface/takbuff/data/LoginData.swift
//
//  LoginData.swift
//  takbuff
//
//  Created by Gurunathan Ramalingam on 18/12/24.
//

import Foundation
import SwiftData

class LoginData {
    private let data: ApiRequest = ApiRequest()
    
    private static let namespace = "auth/"
    
    func auth(
        uData: [String: Any],
        lData: [String: Any],
        completion:@escaping (Dictionary<String, Any>?, Error?) -> ()
    ) {
 
        let reqData: [String: Any] = [
            "user": uData,
            "login": lData
        ]
        
        data.post(path: LoginData.namespace + "login", body: reqData) { tokenData, error in
            guard let tokenData = tokenData else {
                completion(nil, error)
                return
            }
            var tokens = try! JSONDecoder().decode(Dictionary<String, Dictionary<String, String>>.self, from: tokenData)
            var data: Dictionary<String, String> = tokens.removeValue(forKey: "data")!

            let idToken: String = data.removeValue(forKey: "idToken")!
            let accessToken: String = data.removeValue(forKey: "accessToken")!
            
            let idPayload = self.decodeJWTPayload(idToken)
            var userData = idPayload!["user"] as! [String: Any]
            
            if let accessPayload = self.decodeJWTPayload(accessToken) {
                userData["isNew"] = LoginData.isNewToApp(accessPayload: accessPayload)
            }
            
            userData["idToken"] = idToken
            userData["accessToken"] = accessToken
            
            DispatchQueue.main.async {
                completion(userData, nil)
            }
        }
    }
    
    func refresh(
        completion:@escaping (Dictionary<String, Any>?, Error?) -> ()
    ) {

        data.get(path: LoginData.namespace + "refresh_token") { tokenData, error in
            guard let tokenData = tokenData else {
                completion(nil, error)
                return
            }
            
            var tokens = try! JSONDecoder().decode(Dictionary<String, Dictionary<String, String>>.self, from: tokenData)
            
            var data: Dictionary<String, String> = tokens.removeValue(forKey: "data")!
            let idToken = data.removeValue(forKey: "idToken")!
            let accessToken = data.removeValue(forKey: "accessToken")!
            
            let idPayload = self.decodeJWTPayload(idToken)
            var userData = idPayload!["user"] as! [String: Any]
            
            userData["idToken"] = idToken
            userData["accessToken"] = accessToken
            
            DispatchQueue.main.async {
                completion(userData, nil)
            }
        }
    }
    
    func decodeJWTPayload(_ jwt: String) -> [String: Any]? {
        let parts = jwt.components(separatedBy: ".")
        guard parts.count == 3 else {
            print("Invalid JWT format")
            return nil
        }

        let payloadPart = parts[1]

        guard let payloadData = base64UrlDecode(payloadPart) else {
            print("Base64URL decoding failed")
            return nil
        }

        do {
            let jsonObject = try JSONSerialization.jsonObject(with: payloadData, options: [])
            return jsonObject as? [String: Any]
        } catch {
            print("JSON decoding error:", error)
            return nil
        }
    }
    
    private static func isNewToApp(accessPayload: [String: Any]) -> Bool {
        if let accesses = accessPayload["accesses"] as? [[String: String]] {
            for accessValue in accesses {
                for (key, value) in accessValue {
                    if (((key == "object_type") && (value == "U"))){
                        return false
                    }
                }
            }
        }
        return true
    }

    func base64UrlDecode(_ base64Url: String) -> Data? {
        var base64 = base64Url
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")

        let padding = base64.count % 4
        if padding > 0 {
            base64 += String(repeating: "=", count: 4 - padding)
        }

        return Data(base64Encoded: base64)
    }

}
